nansat/mappers/mapper_sentinel1_l1.py

```diff
@@ -415,125 +368,288 @@ def __init__(self, filename, gdalDataset, gdalMetadata,
+        by 2nd order polynomial of input X,Y,Z. Third, this polynomial is used to calculate new
+        Pixel coordinate for ocean surface (Z=0). Fourth, a temporary VRT from original X, Y, Line
+        and corrected Z,Pixel coordinates is created. Fifth, temporary VRT is used for converting
+        original Pixel/Line coordinates into correct Lon/Lat
+
+        Parameters
+        ----------
+        data : dict
+            Original geolocation data from read_geolocation_lut()
+        max_height : int
+            Maximum afordable height
+
+        Returns
+        -------
+        data : dict
+            Corrected geolocation data with new longitude and latitude
+
+        """
+        # don't correct geolocation data if only few points are affected
+        if (data['height'] > max_height).sum() < 10:
+            return data
+
+        # convert XY from degrees to meters in stereographic projection
+        central_point = int(data['shape'][0]/2), int(data['shape'][1]/2)
+        dst_srs = '+proj=stere +datum=WGS84 +ellps=WGS84 +lat_0=%f +lon_0=%f +no_defs' % (
+                        data['latitude'][central_point], data['longitude'][central_point])
+        x, y, z = VRT.transform_coordinates(NSR(), (data['longitude'].flat,
+                                                     data['latitude'].flat,
+                                                     data['height'].flat), NSR(dst_srs))
+        # create training data
+        a = np.vstack([np.ones(x.size), x, x**2, y, y**2, x*y, z, z**2, x*z, y*z]).T
+        # calculate polynomial coefficients for values of Pixel (using least squares)
+        b = np.linalg.lstsq(a, data['pixel'].flat)[0]
+        # pixel_test = np.dot(a, b) # for debugging
+        # set height to zero (ocean surface)
+        a[:, 6:] = 0
+        # calculate Pixel at ocean surface
+        pixel_ocean = np.dot(a, b)
+        high_pixels_idx = data['height'] > max_height
+        tmp_pixel = np.array(data['pixel'])
+        tmp_pixel[high_pixels_idx] = pixel_ocean[high_pixels_idx.flat]
+        new_height = np.zeros(data['height'].shape)
+
+        # create temporary VRT with correct GCPs for converting original pixel/line into lon/lat
+        tmp_gcps = Mapper.create_gcps(x, y, new_height, tmp_pixel, data['line'])
+        tmp_vrt = VRT(data['x_size'], data['y_size'])
+        tmp_vrt.dataset.SetGCPs(tmp_gcps, NSR(dst_srs).wkt)
+        tmp_vrt.tps = True
+        new_lon, new_lat = tmp_vrt.transform_points(data['pixel'].flatten(), data['line'].flatten())
+
+        data['latitude'] = new_lat.reshape(data['shape'])
+        data['longitude'] = new_lon.reshape(data['shape'])
+        data['height'] = new_height
+        return data
+
+    @staticmethod
+    def create_gcps(x, y, z, p, l):
+        """ Create GCPs from geolocation data
+
+        Parameters
+        ----------
+        x, y, z, p, l
+        N-D arrays with vakye of X, Y, Z, Pixel and Line coordinates.
+        X and Y are typically lon, lat.
```

Comment:
what is z?

Reply:
height. It is now added to the docstring.